modules/objdetect/src/qrcode.cpp

```diff
@@ -2566,6 +2587,53 @@ bool QRDecode::versionDefinition()
     return true;
 }
 
+void QRDecode::detectAlignment() {
+    vector<int> alignmentPositions = getAlignmentCoordinates(version);
+    if (alignmentPositions.size() > 0) {
+        // create alignment image
+        static uint8_t alignmentMarker[25] = {
+            0, 0,   0,   0,   0,
+            0, 255, 255, 255, 0,
+            0, 255, 0,   255, 0,
+            0, 255, 255, 255, 0,
+            0, 0,   0,   0,   0
+        };
+        Mat resizedAlignmentMarker(5, 5, CV_8UC1, alignmentMarker);
+        const float module_size = test_perspective_size / version_size;
+        resize(resizedAlignmentMarker, resizedAlignmentMarker,
+               Size(cvRound(module_size * 5.f), cvRound(module_size * 5.f)), 0, 0, INTER_LINEAR_EXACT);
+        threshold(resizedAlignmentMarker, resizedAlignmentMarker, 127., 255, THRESH_BINARY);
+        const float module_offset = 1.9f;
+        const float left_top = cvRound(module_size * (alignmentPositions.back() - 2.f - module_offset)); // add offset
+        const float offset = cvRound(module_size * (5 + module_offset * 2)); // 5 modules in alignment marker, 4 modules in offset
+        Mat subImage(no_border_intermediate, Rect(cvRound(left_top), cvRound(left_top), cvRound(offset), cvRound(offset)));
+        Mat resTemplate;
+        matchTemplate(subImage, resizedAlignmentMarker, resTemplate, TM_SQDIFF_NORMED);
+        double minVal = 0., maxVal = 0.;
+        Point minLoc, maxLoc, matchLoc;
+        minMaxLoc(resTemplate, &minVal, &maxVal, &minLoc, &maxLoc);
+        CV_LOG_INFO(NULL, "Alignment minVal: " << minVal);
+        if (minVal < 0.37) {
+            const float templateOffset = static_cast<float>(resizedAlignmentMarker.size().width)/2.f;
+            Point2f alignmentCoord(Point2f(minLoc.x + left_top + templateOffset, minLoc.y + left_top + templateOffset));
+            alignment_coords = {alignmentCoord};
+            perspectiveTransform(alignment_coords, alignment_coords, homography.inv());
+            CV_LOG_INFO(NULL, "Alignment coords: " << alignment_coords);
+            const float relativePos = (alignmentPositions.back()+0.5f)/version_size;
+
+            vector<Point2f> perspective_points = {{0.f, 0.f}, {test_perspective_size, 0.f},
+                                                  {0.f, test_perspective_size},
+                                                  {relativePos*test_perspective_size, relativePos*test_perspective_size}};
+            vector<Point2f> pts = {original_points[0], original_points[1], original_points[3], alignment_coords[0]};
+            Mat H = findHomography(pts, perspective_points);
+            updatePerspective(H);
```

Comment:
`findHomography` returns empty matrix if homography cannot be estimated. `updatePerspective` result is not checked and we stay with initial estimation. I propose to add it to comments.